core/src/main/java/com/linecorp/armeria/common/metric/MetricKey.java

```diff
@@ -0,0 +1,321 @@
+ * could be annotated with a label {@code "protocol"} so it's counted separately for each protocol type,
+ * such as {@code "protocol=H1C"} and {@code "protocol=H2"}.
+ *
+ * <p>Note that two {@link MetricKey}s with same name and different labels are considered different.
+ */
+public final class MetricKey {
+
+    private final List<String> nameParts;
+    private final Map<MetricLabel, String> labels;
+    private int hashCode;
+
+    /**
+     * Creates a new instance with the specified name parts and with no labels.
+     *
+     * @throws IllegalArgumentException if the name parts 1) is empty or 2) contains an empty string
+     */
+    public MetricKey(String... nameParts) {
+        this(ImmutableList.copyOf(requireNonNull(nameParts, "nameParts")));
+    }
+
+    /**
+     * Creates a new instance with the specified name parts and with no labels.
+     *
+     * @throws IllegalArgumentException if the name parts 1) is empty or 2) contains an empty string
+     */
+    public MetricKey(Iterable<String> nameParts) {
+        this(nameParts, ImmutableMap.of());
+    }
+
+    /**
+     * Creates a new instance with the specified name parts and labels.
+     *
+     * @param nameParts a {@link List} of name parts
+     * @param labels a {@link Map} whose key is a {@link MetricLabel} or a {@link CharSequence}
+     *
+     * @throws IllegalArgumentException if 1) both name parts and labels are empty
+     *                                  2) the name parts contains an empty string or
+     *                                  3) the labels contains an empty name or value
+     */
+    public MetricKey(Iterable<String> nameParts, Map<?, String> labels) {
+        this.nameParts = validateNameParts(nameParts);
+
+        if (labels.isEmpty()) {
+            this.labels = ImmutableMap.of();
+        } else {
+            final ImmutableMap.Builder<MetricLabel, String> builder = ImmutableMap.builder();
+            final Set<String> labelNames = new HashSet<>(labels.size());
+            labels.forEach((k, v) -> {
+                final String name;
+                if (k instanceof MetricLabel) {
+                    name = ((MetricLabel) k).name();
+                    builder.put((MetricLabel) k, v);
+                } else if (k instanceof CharSequence) {
+                    name = k.toString();
+                    builder.put(new MetricLabelImpl(name), v);
+                } else {
+                    throw new IllegalArgumentException(
+                            "labels contains a key of unsupported type: " + k.getClass().getName() +
+                            " (expected: " + MetricLabel.class.getSimpleName() +
+                            " or " + CharSequence.class.getSimpleName() + ')');
+                }
+
+                checkArgument(!name.isEmpty(),
+                              "labels contains an empty label: %s", labels);
+                checkArgument(labelNames.add(name),
```

Comment:
It would also be able to catch `IllegalArgumentException` on `builder.build`, which is thrown when there are duplicate keys (suggested in case it makes sharing code with below easier).